### PR TITLE
[6.x] Restrict concurrency in processing requests (#731)

### PR DESCRIPTION
**Pull request by roncohen**
Backports the following commits to 6.x:
 - Restrict concurrency in processing requests  (#731)

### CODE REVIEW

File: CHANGELOG.asciidoc [hunk 1]
```diff
@@ -16,12 +16,14 @@ https://github.com/elastic/apm-server/compare/4daa36bd5c144cf9182afc62dc8042af66
 - Use type integer instead of number in JSON schemas where applicable {pull}641[641].
 - Set array item types to string in JSON schemas {pull}651[651].
 - Fix issue preventing server from being stopped {pull}704[704].
+- Limit the amount of concurrent requests being processed{pull}731[731].
```
Comment (graphaelli) [inline CHANGELOG.asciidoc]: just noticed we're missing a space before the pr, not a huge deal

File: beater/config.go [hunk 2]
```diff
@@ -107,15 +108,16 @@ func replaceVersion(pattern, version string) string {
...
-		MaxUnzippedSize:    50 * 1024 * 1024, // 50mb
-		MaxHeaderSize:      1 * 1024 * 1024,  // 1mb
-		ConcurrentRequests: 40,
-		ReadTimeout:        2 * time.Second,
-		WriteTimeout:       2 * time.Second,
-		ShutdownTimeout:    5 * time.Second,
-		SecretToken:        "",
-		AugmentEnabled:     true,
+		Host:                "localhost:8200",
+		MaxUnzippedSize:     30 * 1024 * 1024, // 50mb
```
Comment (simitt) [inline beater/config.go]: the comment still states `50mb`

File: beater/handlers.go [hunk 3]
```diff
@@ -105,11 +127,12 @@ func frontendHandler(pf ProcessorFactory, config *Config, report reporter) http.
 		ExcludeFromGrouping: regexp.MustCompile(config.Frontend.ExcludeFromGrouping),
 	}
 	return logHandler(
-		killSwitchHandler(config.Frontend.isEnabled(),
-			ipRateLimitHandler(config.Frontend.RateLimit,
-				corsHandler(config.Frontend.AllowOrigins,
-					processRequestHandler(pf, &prConfig, report,
-						decoder.DecodeUserData(decoder.DecodeLimitJSONData(config.MaxUnzippedSize), config.AugmentEnabled))))))
+		concurrencyLimitHandler(config,
+			killSwitchHandler(config.Frontend.isEnabled(),
```
Comment (simitt) [inline beater/handlers.go]: Sorry for the late comment, but what do you think about changing this order? The `killSwitchHandler` only checks a config boolean, so it is a cheap operation. If we first run the `concurrencyLimitHandler` then you could fill up the concurrency queue with requests to an endpoint that might not even be enabled.

**Comment by roncohen** [comment]
this already got merged into `master`, so i'll merge this and create a follow up PR instead

**Comment by roncohen** [comment]
Follow up: https://github.com/elastic/apm-server/pull/738

**Pull request merged**